### ISSUE TITLE
Creating an app, then changing to a plugin in the pubspec leaves stale top-level platform files

### ISSUE THREAD

**Issue by tanlianghao** [flutter/flutter#102029]
## Steps to Reproduce

1. Create a new project using flutter create
2. Add the plugin field to the pubspec.yaml file 
3. When the flutter version is 2.2.0，Failed to generate Generated xconfig file when the flutter pub get is executed when the plugin field exists
4. The generation xconfig file can be generated normally if Flutter 2.0.4 is used

**Expected results:** 
Flutter pub get generates files normally

**Actual results:** 
No Generated xconfig  files
<details>
<summary>Code sample</summary>

```
/// pubspec.yaml
name: test
description: A new Flutter project.

publish_to: 'none' # Remove this line if you wish to publish to pub.dev
version: 1.0.0+1

environment:
  sdk: ">=2.12.0 <3.0.0"

dependencies:
  flutter:
    sdk: flutter
  flutter_web_plugins:
    sdk: flutter
  cupertino_icons: ^1.0.2
dev_dependencies:
  flutter_test:
    sdk: flutter

flutter:
  plugin:
    platforms:
      web:
        pluginsClass: WebRegistrant
        fileName: web_register.dart
```

```
///  web_register.dart
import 'package:flutter_web_plugins/flutter_web_plugins.dart';

class WebRegistrant {
  static void registerWith(Registrar registrar) {
    print('------test');
  }
}

```

</details>

<details>
  <summary>flutter version info</summary>

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 2.2.0, on macOS 12.3.1 21E258
    darwin-x64, locale zh-Hans-CN)
[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from:
      https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the
      Android SDK components.
      (or visit
      https://flutter.dev/docs/get-started/install/macos#androi
      d-setup for detailed instructions).
      If the Android SDK has been installed to a custom
      location, please use
      `flutter config --android-sdk` to update to that
      location.

[✓] Xcode - develop for iOS and macOS
[✓] Chrome - develop for the web
[!] Android Studio (not installed)
[✓] VS Code (version 1.66.2)
[✓] Connected device (1 available)

! Doctor found issues in 2 categories.
```

</details>

**Comment by darshankawar** [comment]
> Channel stable, 2.2.0,

You are on old version of stable. Current version is 2.10.4. I recommend you to upgrade to latest and try again and see if you still get same behavior or not. If yes, write in comments and I'll reopen it. Closing for now.

**Comment by tanlianghao** [comment]
> > Channel stable, 2.2.0,
> 
> You are on old version of stable. Current version is 2.10.4. I recommend you to upgrade to latest and try again and see if you still get same behavior or not. If yes, write in comments and I'll reopen it. Closing for now.

<details>
<summary>flutter version info</summary>



```Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 2.10.4, on macOS 11.4 20F71 darwin-x64,
    locale zh-Hans-CN)
[✓] Android toolchain - develop for Android devices (Android SDK version
    32.0.0)
[!] Xcode - develop for iOS and macOS (Xcode 12.5.1)
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.2)
[✓] VS Code (version 1.66.1)
[✓] Connected device (2 available)
[✓] HTTP Host Availability
```
</details>
With flutter 2.10.4, the Generated.xcconfig file cannot be generated when flutter pub get is executed.
I wonder why 2.0.4 works, but not the others

**Comment by darshankawar** [comment]
> flutter:
>   plugin:
>     platforms:
>       web:
>         pluginsClass: WebRegistrant
>         fileName: web_register.dart

Why would you want to add plugin like this ? Can you directly try to add as you did below and check if that works ?

```
///  web_register.dart
import 'package:flutter_web_plugins/flutter_web_plugins.dart';

class WebRegistrant {
  static void registerWith(Registrar registrar) {
    print('------test');
  }
}

```

**Comment by tanlianghao** [comment]
> > flutter:
> > plugin:
> > platforms:
> > web:
> > pluginsClass: WebRegistrant
> > fileName: web_register.dart
> 
> Why would you want to add plugin like this ? Can you directly try to add as you did below and check if that works ?
> 
> ```
> ///  web_register.dart
> import 'package:flutter_web_plugins/flutter_web_plugins.dart';
> 
> class WebRegistrant {
>   static void registerWith(Registrar registrar) {
>     print('------test');
>   }
> }
> ```
Business implementation. The directory in the project includes mobile terminal and Web terminal. Some business implementations need to load different business codes according to different platforms.

From the previous situation, this is ok, after all, version 2.0.4 is supported. However, after the upgrade of FLUTTER, there began to be problems. From the current situation, it is the higher version that does not support this use, but I can't find any documentation about this change.

**Comment by tanlianghao** [comment]
@darshankawar If you can't help solve it, please reopen the issue

**Comment by tanlianghao** [comment]
I can't find any documentation to explain why Flutter 2.0.4 works but not after the upgrade

**Comment by Chouyoug05** [comment]
Yes

**Comment by darshankawar** [comment]
Thanks for feedback. Reopening.

**Comment by darshankawar** [comment]
@tanlianghao 
Since the said file falls under iOS part, when you add a plugin, can you try running `pod install` and then check if the said file is generated or not ?

**Comment by tanlianghao** [comment]
> @tanlianghao Since the said file falls under iOS part, when you add a plugin, can you try running `pod install` and then check if the said file is generated or not ?

No, you must have the Generated. Xcconfig file before pod install

**Comment by tanlianghao** [comment]
This exception can easily be repeated by creating a new project like the code I provided above

**Comment by darshankawar** [comment]
@tanlianghao 
Does the same behavior occur on latest master too ?

`flutter channel master`
`flutter upgrade`

**Comment by tanlianghao** [comment]
> `flutter channel master`

I tried, but it didn't work
<details>
<summary>flutter doctor</summary>

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel master, 2.13.0-0.0.pre.653, on macOS 11.4 20F71
    darwin-x64, locale zh-Hans-CN)
[✓] Android toolchain - develop for Android devices (Android SDK
    version 32.0.0)
[!] Xcode - develop for iOS and macOS (Xcode 12.5.1)
    ✗ Flutter requires Xcode 13 or higher.
      Download the latest version or update via the Mac App Store.
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.2)
[✓] VS Code (version 1.66.2)
[✓] Connected device (1 available)
[✓] HTTP Host Availability
```

</details>

<details>
<summary>flutter version</summary>

```
Flutter 2.13.0-0.0.pre.653 • channel master •
https://github.com/flutter/flutter.git
Framework • revision 3fefb2d5eb (5 minutes ago) • 2022-04-20 22:34:07
-0700
Engine • revision 4274987f68
Tools • Dart 2.18.0 (build 2.18.0-37.0.dev) • DevTools 2.12.2
```

</details>

**Comment by darshankawar** [comment]
Thanks for the update. I tried the same on both versions and don't see the said file being generated. Having said that, I am not sure if this is intended or not, so will keep this issue open and label for further insights from the team on expected behavior.


<details>
<summary> stable, master flutter doctor -v </summary>

```

[✓] Flutter (Channel stable, 2.10.5, on macOS 12.2.1 21D62 darwin-x64, locale
    en-GB)
    • Flutter version 2.10.5 at /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 5464c5bac7 (12 hours ago), 2022-04-18 09:55:37 -0700
    • Engine revision 57d3bac3dd
    • Dart version 2.16.2
    • DevTools version 2.9.2

[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.

[✓] Flutter (Channel master, 2.13.0-0.0.pre.653, on macOS 12.2.1 21D62
    darwin-x64, locale en-GB)
    • Flutter version 2.13.0-0.0.pre.653 at
      /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 3fefb2d5eb (15 minutes ago), 2022-04-20 22:34:07 -0700
    • Engine revision 4274987f68
    • Dart version 2.18.0 (build 2.18.0-37.0.dev)
    • DevTools version 2.12.2
    
[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.



```

</details>

**Comment by jmagman** [comment]
> Add the plugin field to the pubspec.yaml file

```
flutter:
  plugin:
    platforms:
      web:
        pluginsClass: WebRegistrant
        fileName: web_register.dart
```

Creating a normal flutter app creates the `ios` directory, but then you added the `plugin` key which tells Flutter tooling that it's not an app, [it's a plugin](https://docs.flutter.dev/development/packages-and-plugins/developing-packages#types).  If you are developing a plugin there is no `ios` directory, there's an `example/ios` directory that has the files you're probably expecting.  

In any case I would start over and make your project as a plugin from the start:
```
$ flutter create -t plugin my_plugin
```
or if you are actually trying to make and ship an app, that isn't how you add a plugin.  I didn't quite follow "The directory in the project includes mobile terminal and Web terminal. Some business implementations need to load different business codes according to different platforms" but if you need to use different logic on different platforms you don't make a plugin, you should instead check in your code what platform you're on, and call different logic accordingly, or import different libraries accordingly.

I guess I'll keep this open since there's nothing stopping users from adding the `plugin` key, at which point the stale app files like `ios` are just hanging out, the tool could warn you you're in this state?

**Comment by tanlianghao** [comment]
> > Add the plugin field to the pubspec.yaml file
> 
> ```
> flutter:
>   plugin:
>     platforms:
>       web:
>         pluginsClass: WebRegistrant
>         fileName: web_register.dart
> ```
> 
> Creating a normal flutter app creates the `ios` directory, but then you added the `plugin` key which tells Flutter tooling that it's not an app, [it's a plugin](https://docs.flutter.dev/development/packages-and-plugins/developing-packages#types). If you are developing a plugin there is no `ios` directory, there's an `example/ios` directory that has the files you're probably expecting.
> 
> In any case I would start over and make your project as a plugin from the start:
> 
> ```
> $ flutter create -t plugin my_plugin
> ```
> 
> or if you are actually trying to make and ship an app, that isn't how you add a plugin. I didn't quite follow "The directory in the project includes mobile terminal and Web terminal. Some business implementations need to load different business codes according to different platforms" but if you need to use different logic on different platforms you don't make a plugin, you should instead check in your code what platform you're on, and call different logic accordingly, or import different libraries accordingly.
> 
> I guess I'll keep this open since there's nothing stopping users from adding the `plugin` key, at which point the stale app files like `ios` are just hanging out, the tool could warn you you're in this state?

Yes, because my project is currently using flutter2.0.4, this version can get the results I expected by adding the plugin field to pubspec.yaml. 
Recently, I was thinking about upgrading the flutter version of the project to the latest, so there was the problem I described above. 
What I wonder now is why the version before 2.2.0 is OK, but the later version is not. Did you fix it on purpose? 
I want to know a definite answer. Is this a question or a normal performance in the new version?

**Comment by jmagman** [comment]
> Is this a question or a normal performance in the new version?

You're exploiting a bug that was maybe fixed in #76406.  In any case you're not using that pubspec key correctly, it doesn't sound like you're trying to build a plugin, which is what adding that key indicates.  Whatever version you're on, that's not the correct way to detect the platform.

**Comment by jmagman** [comment]
This is a duplicate of https://github.com/flutter/flutter/issues/78785
